functions/post_data_test.go

```diff
@@ -0,0 +1,260 @@
+package functions
+
+import (
+	"errors"
+	"testing"
+)
+
+const testURL = "http://httpbin.org/post"
+
+func TestPostData(t *testing.T) {
+	var tests = []struct {
+		existingHeaders map[string]string
+		args            []string
+		expected        string
+		err             error
+	}{
+		{
+			map[string]string{},
+			[]string{},
+			"",
+			errURLMissing,
+		},
+		{
+			map[string]string{},
+			[]string{"", "foo", "bar"},
+			"",
+			errURLEmpty,
+		},
+		{
+			map[string]string{},
+			[]string{testURL, "url"},
+			testURL,
+			nil,
+		},
+		{
+			map[string]string{},
+			[]string{testURL, "foo"},
+			"",
+			errors.New("error: element foo not found in response body"),
+		},
+		{
+			map[string]string{},
+			[]string{testURL, "form", "data1=123", "data2=456", "data3=789"},
+			`{"data1":"123","data2":"456","data3":"789"}`,
+			nil,
+		},
+		{
+			map[string]string{
+				"x-previous-header1": "foo=bar",
+				"x-previous-header2": "abc=xyz",
+			},
+			[]string{testURL, "form", "data1=123", "x-previous-header2", "data2=456"},
+			`{"abc":"xyz","data1":"123","data2":"456"}`, // this API will return the keys in alphabetical order
+			nil,
+		},
+	}
+
+	for _, tc := range tests {
+		actual, err := PostData(tc.existingHeaders, tc.args)
+		if actual != tc.expected {
+			t.Errorf("PostData(%v, %v): expected %v, actual %v", tc.existingHeaders, tc.args, tc.expected, actual)
+		}
+		if err != tc.err && err.Error() != tc.err.Error() {
+			t.Errorf("PostData(%v, %v): expected %v, got: %v", tc.existingHeaders, tc.args, tc.err, err)
+		}
+	}
+}
+
+func TestArgsToURLResponseElement(t *testing.T) {
+	var tests = []struct {
+		args                    []string
+		expectedURL             string
+		expectedResponseElement string
+		err                     error
+	}{
+		{
+			[]string{"foo", "bar"},
+			"foo",
+			"bar",
+			nil,
+		},
+		{
+			[]string{"foo", "bar", "abc", "xyz"},
+			"foo",
+			"bar",
+			nil,
+		},
+		{
+			[]string{"", "bar", "abc", "xyz"},
+			"",
+			"",
+			errURLEmpty,
+		},
+		{
+			[]string{"foo", "", "abc", "xyz"},
+			"foo",
+			"",
+			nil,
+		},
+	}
+
+	for _, tc := range tests {
+		actualURL, actualResponseElement, err := argsToURLResponseElement(tc.args)
+		if actualURL != tc.expectedURL {
+			t.Errorf("argsToURLResponseElement(%v): expected URL %v, actual URL %v", tc.args, tc.expectedURL, actualURL)
+		}
+		if actualResponseElement != tc.expectedResponseElement {
+			t.Errorf("argsToURLResponseElement(%v): expected response element %v, actual response element %v", tc.args, tc.expectedResponseElement, actualResponseElement)
+		}
+		if err != tc.err && err.Error() != tc.err.Error() {
+			t.Errorf("argsToURLResponseElement(%v): expected %v, got: %v", tc.args, tc.err, err)
+		}
+	}
+}
+
+func TestArgsToRequestBody(t *testing.T) {
```

Comment:
I think this test can be simplified if using `url.Values`.

Reply:
Using `url.Values` now